url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/beans/StandardBeanInfo.java

```diff
@@ -0,0 +1,1505 @@
+                table.put(STR_INDEXED, STR_VALID);
+                table.put(STR_INDEXED + PREFIX_GET, indexedGetter);
+                table.put(STR_INDEXED + PREFIX_SET, indexedSetter);
+                table.put(STR_INDEXED + STR_PROPERTY_TYPE, indexedPropType);
+                continue;
+            }
+
+            // default rule - invalid property
+            table.put(STR_NORMAL, STR_INVALID);
+            table.put(STR_INDEXED, STR_INVALID);            
+        }
+
+    }
+
+    /**
+     * Introspects the supplied Bean class and returns a list of the Events of
+     * the class
+     * 
+     * @return the events
+     * @throws IntrospectionException
+     */
+    @SuppressWarnings({ "rawtypes" })
+    private EventSetDescriptor[] introspectEvents() throws IntrospectionException {
+        // Get descriptors for the public methods
+        // FIXME: performance
+        MethodDescriptor[] theMethods = introspectMethods();
+
+        if (theMethods == null)
+            return null;
+
+        HashMap<String, HashMap> eventTable = new HashMap<String, HashMap>(
+                theMethods.length);
+
+        // Search for methods that add an Event Listener
+        for (int i = 0; i < theMethods.length; i++) {
+            introspectListenerMethods(PREFIX_ADD, theMethods[i].getMethod(),
+                    eventTable);
+            introspectListenerMethods(PREFIX_REMOVE, theMethods[i].getMethod(),
+                    eventTable);
+            introspectGetListenerMethods(theMethods[i].getMethod(), eventTable);
+        }
+
+        ArrayList<EventSetDescriptor> eventList = new ArrayList<EventSetDescriptor>();
+        for (Map.Entry<String, HashMap> entry : eventTable.entrySet()) {
+            HashMap table = entry.getValue();
+            Method add = (Method) table.get(PREFIX_ADD);
+            Method remove = (Method) table.get(PREFIX_REMOVE);
+
+            if ((add == null) || (remove == null)) {
+                continue;
+            }
+
+            Method get = (Method) table.get(PREFIX_GET);
+            Class<?> listenerType = (Class) table.get("listenerType"); //$NON-NLS-1$
+            Method[] listenerMethods = (Method[]) table.get("listenerMethods"); //$NON-NLS-1$
+            EventSetDescriptor eventSetDescriptor = new EventSetDescriptor(
+                    decapitalize(entry.getKey()), listenerType, listenerMethods, add,
+                    remove, get);
+
+            eventSetDescriptor.setUnicast(table.get("isUnicast") != null); //$NON-NLS-1$
+            eventList.add(eventSetDescriptor);
+        }
+
+        EventSetDescriptor[] theEvents = new EventSetDescriptor[eventList
+                .size()];
+        eventList.toArray(theEvents);
+
+        return theEvents;
+    }
+
+    /* MODIFIED FOR THE MSGPACK PROJECT
+     * find the add, remove listener method
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private static void introspectListenerMethods(String type,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1376-1403 into a separate method.